fetcher/fetcher.go

```diff
@@ -222,36 +198,44 @@ func (bot *fetcherImpl) Run() {
 	}
 }
 
-// SetContext is part of the Fetcher interface
-func (bot *fetcherImpl) SetContext(ctx context.Context) {
-	bot.ctx = ctx
+// Run is part of the Fetcher interface
+func (bot *fetcherImpl) Run(ctx context.Context) error {
+	// In theory a buffer of size one should be enough, but let's make it bigger.
+	bot.blockQueue = make(chan *rpcs.EncodedBlockCert, 5)
+
+	ctx, cancelFunc := context.WithCancel(ctx)
+	defer cancelFunc()
+
+	ch0 := make(chan error)
+	go func() {
+		ch0 <- bot.processQueue(ctx)
+	}()
+
+	ch1 := make(chan error)
+	go func() {
+		ch1 <- bot.mainLoop(ctx)
+	}()
+
+	select {
+	case err := <-ch0:
+		cancelFunc()
+		<-ch1
```

Comment:
Why do we need to wait for the other goroutine to finish before exiting? I think you can just return the errors.

Reply:
I guess here we don't. In the other case we do because we want to finish database operations before closing the database.